src/sys/windows/selector.rs

```diff
@@ -407,13 +407,50 @@ impl Selector {
 pub struct SelectorInner {
     cp: Arc<CompletionPort>,
     update_queue: Mutex<VecDeque<Arc<Mutex<SockState>>>>,
+    pending_queue: Mutex<HashMap<u64, Arc<Mutex<SockState>>>>,
```

Comment:
I'm concerned about using user_data as the key into the HashMap. If the user token is re-used between `deregister()` and `register()` of a different socket, the internal state gets messed up.

Reply:
Ideally `pending_queue` and `update_queue` would be behind the same Mutex. 
I didn't get around to doing that in #1149 but it should be done eventually nonetheless.

Reply:
pending_queue is a bit of a misnomer here. This is not a queue of pending socket, but rather one of *all* sockets (regardless of whether they're pending/idle/cancelled/deleted).

Reply:
It looks like explicitly deregistered sockets are removed from this queue, but implicitly deregistered ones are. Maybe make that more consistent?

Reply:
@piscisaureus 

> I'm concerned about using user_data as the key into the HashMap. If the user token is re-used between `deregister()` and `register()` of a different socket, the internal state gets messed up.

I agree, not a good idea to use token as key.

> Ideally `pending_queue` and `update_queue` would be behind the same Mutex.
Ok.

> "pending_queue" is a bit of a misnomer here. 

I agree, need a better name here.

> It looks like explicitly deregistered sockets are removed from this queue, but implicitly deregistered ones are. Maybe make that more consistent?

What do you mean by "implicitly deregistered sockets"? Could you give an example of such a case? Thanks